README
MemcachedPP      - pure Perl lightweight implementation of Memcached
MemcachedPP::Web - simple web interface to MemcachedPP

USAGE

See ./script/memcachedpp -h for instructions.

Example1:

./script/memcachedpp -l 127.0.0.1:9191 -T 60 -f sqlite.db -d -p memdpp.pid -D -L memdpp.log

Meaning:
* run memcachedpp as daemon write PID to memdpp.pid
* listen to 127.0.0.1:9191 for incomming connections,
  disconnect after 60 seconds of idling or transport delay;
* setup SQLite storage within database file of sqlite.db
* write logs to memdpp.log, set log level to DEBUG

Example2:

./script/memcachedpp -l somesocket -D

Meaning:
* run memcachedpp in foreground mode
* listen to UNIX socket 'somesocket' for incomming connections;
* setup storage in memory
* write logs to screen, set log level to DEBUG

Example3:

./script/memcachedpp -S -p memdpp.pid

Meaning:
* read memcachedpp PID from file memdpp.pid;
* stop memcachedpp daemon;


MemcachedPP::Web fairly uses Cache::Memcached client to talk to
the MemcachedPP backend.

There are two ways to run the stuff:

* run the backend server first, then run the web server

./script/memcachedpp -l 127.0.0.1:9191 -f memcachedpp.db -d -p memcachedpp.pid
./script/webdemo daemon

* OR let the web server fork a separate backend process

WEBDEMO_WITH_BACKEND=1 ./script/webdemo daemon

Both ways will set memcachedpp up with SQLite storage in file called
memcachedpp.db. Web service will be available at http://*:3000.


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

Alternatively, to install with Module::Build, you can use the following commands:

	perl Build.PL
	./Build
	./Build test
	./Build install


DEPENDENCIES

Log::Dispatch - for comprehensive logging
Cache::Memcached - to talk to MemcachedPP service
Mojolicious - to produce web interface
AnyEvent, EV - for asynchronous server
DBI, DBD::SQLite - for storage


COPYRIGHT AND LICENCE

Copyright (C) 2010, Alexey Skorikov

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for MemcachedPP

0.0.1  Mon Aug  9 02:44:51 2010
       Initial release.



Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME                => 'MemcachedPP',
    AUTHOR              => 'Alexey Skorikov <alexey@skorikov.name>',
    VERSION_FROM        => 'lib/MemcachedPP.pm',
    ABSTRACT_FROM       => 'lib/MemcachedPP.pm',
    PL_FILES            => {},
    PREREQ_PM => {
        'Log::Dispatch'       => 2.26,
        'Cache::Memcached'    => 1.29,
        'Mojolicious'         => 0.9999,
        'AnyEvent'            => 5.27,
        'EV'                  => 3.9,
        'DBI'                 => 1.61,
        'DBD::SQLite'         => 1.29,
        'Test::More'          => 0,
        'Test::Pod'           => 1.14,
        'File::Temp'          => 0,
        'version'             => 0,
    },
    dist                => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean               => { FILES => 'MemcachedPP-*' },
);


lib/MemcachedPP.pm
package MemcachedPP;

use strict;

use Getopt::Std;
use Fcntl qw(:DEFAULT :flock);
use POSIX qw(setsid setuid setpgid);
use Log::Dispatch;

use MemcachedPP::Log;
use MemcachedPP::Server;

use constant FORK_WAIT => 2;
use constant FORK_MAX_ATTEMPTS => 10;

our $VERSION = '0.86';


sub daemonize
{
    my ($class, $pid_file, $uid, $gid) = @_;

    print "Starting memcachedpp... ";

    if (my $old_pid = $class->get_pid($pid_file)) {
        kill(0, $old_pid) and
            die "The daemon is already running with PID $old_pid";
    }

    sysopen(PIDFILE, $pid_file, O_RDWR|O_CREAT) or
        die "Can't open PID file $pid_file: $!";
    flock(PIDFILE, LOCK_EX | LOCK_NB) or
        die "PID file $pid_file is already locked";
    sysseek PIDFILE, 0, 0 and truncate PIDFILE, 0 or
        die "PID file $pid_file is not writable: $!";

    my $attempt = 0;
    my $pid;

    while (not defined ($pid = fork())) {
        die "Too many failed fork attempts: $!\n"
            if ++$attempt > FORK_MAX_ATTEMPTS;
        warn "Fork failed: $!\n";
        sleep FORK_WAIT;
    }

    if ($pid)
    {
        syswrite PIDFILE, "$pid\n", length("$pid\n") and close(PIDFILE)
            or die "Can't write PID file $pid_file: $!";

        print "Success (PID=$pid)\n";
        exit 1;
    }

    umask 0;
    POSIX::setsid()                 or die "Can't start a new session: $!";
    $uid and POSIX::setuid($uid)    || die "Can't set UID: $!";
    $gid and POSIX::setpgid($gid)   || die "Can't set GID: $!";
    open STDIN,  q{<}, '/dev/null'  or die "Can't read /dev/null: $!";
    open STDOUT, q{+>&STDIN}        or die "Can't write to STDIN: $!";
    open STDERR, q{+>&STDIN}        or die "Can't write to STDIN: $!";
    $0 = 'memcachedpp';

    1;
}

sub get_pid
{
    my ($class, $pid_file) = @_;

    (-r $pid_file) or return 0;
    open(PIDFILE, '<', $pid_file) or return 0;
    my $pid = <PIDFILE>;
    close(PIDFILE);
    $pid or return 0;

    return int($pid);
}

sub start {
    my ($class, %opts) = @_;
    %opts or getopts('l:f:p:u:g:dhKT:L:DS', \%opts);

    $opts{h}                 && $class->print_usage;
    my $pid_file = $opts{p} || 'memcachedpp.pid';

    if ($opts{S}) {
        $class->stop_daemon($pid_file);
        exit;
    }

    my $listen    = $opts{l} || $class->print_usage;
    my $db_file   = $opts{f};
    my $daemonize = $opts{d};
    my $log_level = $opts{D} ? 'debug' : 'warning';

    my $logger = MemcachedPP::Log::set_logger( Log::Dispatch->new );

    if ($daemonize && ( my $log_file = $opts{L} ) )
    {
        require Log::Dispatch::File;
        $logger->add( Log::Dispatch::File->new
            ( name      => 'logfile',
              min_level => $log_level,
              filename  => $log_file,
              mode      => '>>',
              newline   => 1,
            )
        );
    } else
    {
        require Log::Dispatch::Screen;
        $logger->add( Log::Dispatch::Screen->new
            ( name      => 'screen',
              min_level => $log_level,
              newline   => 1,
            )
        );
    }

    local $SIG{__WARN__} = sub { WARN $_[0] };
    local $SIG{__DIE__} = sub { CRITICAL $_[0] };

    my $storage;

    if ($db_file) {
        require MemcachedPP::Storage::SQLite;
        $storage = MemcachedPP::Storage::SQLite->new(dbfile => $db_file);
    } else {
        require MemcachedPP::Storage;
        $storage = MemcachedPP::Storage->new;
    }

    my %args = (
        listen    => [ $listen ],
        storage   => $storage,
        keepalive => $opts{K},
    );
    $args{timeout} = $opts{T} if defined $opts{T};

    my $server = MemcachedPP::Server->new(%args);

    if ($daemonize) {
        my ($uid, $guid) = ($opts{u}, $opts{g});
        $class->daemonize($pid_file, $uid, $guid);
    }

    $server->run;
}

sub stop_daemon
{
    my ($class, $pid_file) = @_;

    print "Stopping memcachedpp... ";

    my $pid = $class->get_pid($pid_file) or die
        "The daemon is probably not running. Failed to read PID file $pid_file";

    if ( kill(0, $pid) ) {
        kill(15, $pid) or kill(9, $pid) or die "Failed to kill the process $pid";
    } else {
        print "Failed to locate the process with PID $pid. " .
              "Probably it is not running. Removing PID file... ";
    }

    unlink $pid_file or die "Failed to remove PID file $pid_file";
    print "Success\n";
    1;
}

sub print_usage
{
    print <<EOF_USAGE;
Usage:

  Start:
    $0 -l <host:port> [-K] [-T <timeout>] [-f <db_file>] [-d [-p <pid_file>] [-u <UID>] [-g <GID>] ] [-L <log_file>] [-D]

  Stop:
    $0 -S [-p <pid_file>]

EOF_USAGE
    exit;
}

1;
__END__


=head1 NAME

MemcachedPP - pure Perl lightweight implementation of Memcached


=head1 VERSION

This document describes MemcachedPP version 0.86


=head1 SYNOPSIS

    use MemcachedPP;
    use MemcachedPP::Storage;

    my $server = MemcachedPP::Server->new(
        host    => '127.0.0.1',
        port    => 9191,
        storage => MemcachedPP::Storage->new(),
    );

    $server->run;


=head1 INTERFACE

Example 1:

./memcachedpp -l 127.0.0.1:9191 -T 60 -f sqlite.db -d -p memdpp.pid -D -L memdpp.log

Meaning:
* run memcachedpp as daemon write PID to memdpp.pid
* listen to 127.0.0.1:9191 for incomming connections,
  disconnect after 60 seconds of idling or transport delay;
* setup SQLite storage within database file of sqlite.db
* write logs to memdpp.log, set log level to DEBUG

Example 2:

./memcachedpp -l somesocket -D

Meaning:
* run memcachedpp in foreground mode
* listen to UNIX socket 'somesocket' for incomming connections;
* setup storage in memory
* write logs to screen, set log level to DEBUG

Example 3:

./memcachedpp -S -p memdpp.pid

Meaning:
* read memcachedpp PID from file memdpp.pid;
* stop memcachedpp daemon;


=head1 DEPENDENCIES

Log::Dispatch - for comprehensive logging
Cache::Memcached - to talk to MemcachedPP service
Mojolicious - to produce web interface
AnyEvent, EV - for asynchronous server
DBI, DBD::SQLite - for storage


=head1 AUTHOR

Alexey Skorikov  C<< <alexey@skorikov.name> >>


=head1 LICENCE AND COPYRIGHT

Copyright (c) 2010, Alexey Skorikov C<< <alexey@skorikov.name> >>. All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.


lib/MemcachedPP/Log.pm
package MemcachedPP::Log;

use strict;
use POSIX qw(strftime);

our $LOGGER;
our %IMPORT_CALLED;


sub import
{
    my $class = shift;

    no strict qw(refs);

    my $caller_pkg = caller();

    return 1 if $IMPORT_CALLED{$caller_pkg}++;

    for(qw(DEBUG INFO NOTICE WARNING ERROR CRITICAL ALERT EMERGENCY)) {
        my $sub_name = $_;
        my $level    = lc($_);
        *{"$caller_pkg\::$sub_name"} = sub {
            my $message = strftime("[%d %b %Y %H:%M:%S]", localtime) .
                          " [$level] " . $_[0];
            $LOGGER ?
                $LOGGER->log(
                    level => $level,
                    message => $message
                ) :
                print STDOUT $message . "\n";
        };
    }
}

sub set_logger
{
    my $logger = shift;
    return $LOGGER = $logger if ((ref $logger) =~ /^Log::Dispatch/);
    return 0;
}

1;
__END__

lib/MemcachedPP/Web.pm
package MemcachedPP::Web;

use strict;
use warnings;

use Mojolicious 0.9999;
use Mojolicious::Lite;

use URI::Escape;
use Cache::Memcached;
$Cache::Memcached::HAVE_ZLIB = 0;

our $VERSION = '0.86';

app->home->parse( $ENV{WEBDEMO_HOME} );
app->renderer->root( app->home->rel_dir('templates') );
app->static->root( app->home->rel_dir('public') );
app->renderer->default_handler('ep');

get  '/'         => \&list_items => 'list';
get  '/list'     => \&list_items => 'list';
get  '/get/:key' => \&get_item   => 'get';
get  '/set/'     =>              => 'set';
get  '/set/:key' => \&get_item   => 'set';
post '/set'      => \&set_item   => 'list';
post '/del'      => \&del_items  => 'list';
get  '/del/:key' => \&del_item   => 'list';

sub list_items {
    my $c = shift;
    my $page_limit = 20;
    my @items;
    my $cachedump_str = "cachedump 1 $page_limit";
    my $stats = app->memd->stats([ $cachedump_str, 'items' ])
        ->{hosts}{$ENV{WEBDEMO_MEMCACHED}};
    my $stats_cachedump = $stats->{$cachedump_str};
    my $stats_items = $stats->{items};
    my $now = time();

    foreach my $line (split /^/, $stats_cachedump) {
        if (
            my ($key, $length, $exptime) = 
                ($line =~
                    m/^ITEM\s(\p{IsGraph}+)\s+\[\s*(\d{1,11})\s+b\;\s+(\d{1,11})\s+s\s*\]\s*$/gm)
        ) {
            my $ttl = ($exptime > 0) ? int($exptime - $now) : 0;
            my $value = substr(app->memd->get($key), 0, 128);
            push @items, [ uri_escape($key), $key, $length, $ttl, $value ];
        }
    }

    $c->stash(items => \@items);
    $c->render;
}

sub del_items
{
    my $c = shift;
    my $params = $c->req->body_params->to_hash;

    foreach my $key (keys %$params) {
        app->memd->delete($key) if ('on' eq $params->{$key});
    }

    list_items($c);
}

sub del_item
{
    my $c = shift;

    app->memd->delete( $c->param('key') );
    list_items($c);
}

sub get_item
{
    my $c = shift;
    my $item_key = $c->param('key');
    my $item_value = app->memd->get( $item_key );
    $c->render_not_found unless defined $item_value;

    $c->stash(item =>
        [ uri_escape($item_key), $item_key, length($item_value), $item_value ]
    );
    $c->render;
}

sub set_item
{
    my $c = shift;
    my $params = $c->req->body_params->to_hash;
    my $warnings = $c->stash('warnings');

    if (defined $params->{key} && ($params->{key} =~ m/[^\/\.\*]+/)) {
        $params->{key}     = substr($params->{key},   0, 250);
        $params->{value}   = substr($params->{value}, 0, 1048576); # 1 Mb
        $params->{exptime} = int($params->{exptime} || 0);
        app->memd->set($params->{key}, $params->{value}, $params->{exptime})
            or push @$warnings, "Set operation has failed";
    } else {
        push @$warnings, "Illegal key specified";
    }

    list_items($c);
}

app->plugins->add_hook(
    before_dispatch => sub {
        my ($self, $c) = @_;

        $c->stash(template_class => __PACKAGE__);

        $c->stash(warnings => []);
    }
);

(ref app)->attr('memd');
app->memd(
    new Cache::Memcached {
        servers            => [ $ENV{WEBDEMO_MEMCACHED} ],
        compress_threshold => 0,
    }
);

app->start;
__END__

lib/MemcachedPP/Server.pm
package MemcachedPP::Server;

use strict;
use warnings;

use Socket qw(IPPROTO_TCP TCP_NODELAY);

use EV;
use AnyEvent;
use AnyEvent::Handle;
use AnyEvent::Socket;

use MemcachedPP::Log;

our $VERSION = '0.86';


sub new {
    my($class, @args) = @_;

    my $self = bless {
        no_delay  => 0,
        timeout   => 3600,
        keepalive => 1,
        @args,
    }, $class;

    ref $self->{storage} or
        die "Invalid or unspecified storage\n";

    return $self;
}

sub storage { shift->{storage} }

sub start_listen {
    my $self = shift;
    my @listen = @{$self->{listen} || [ ($self->{host} || '') . ":$self->{port}" ]};
    for my $listen (@listen) {
        push @{$self->{listen_guards}}, $self->_create_tcp_server($listen);
    }
}

sub register_service {
    my $self = shift;
    DEBUG "Registering service";

    $self->start_listen;

    $self->{exit_guard} = AE::cv {
        delete $self->{listen_guards};
    };
    $self->{exit_guard}->begin;
}

sub run {
    my $self = shift;
    $self->register_service;

    my $w; $w = AE::signal QUIT => sub { $self->{exit_guard}->end; undef $w };
    $self->{exit_guard}->recv;

    DEBUG "Shutting down";
}

sub _create_tcp_server {
    my ( $self, $listen ) = @_;

    my($host, $port, $is_tcp);
    if ($listen =~ /:\d+$/) {
        ($host, $port) = split /:/, $listen;
        $host = undef if $host eq '';
        $is_tcp = 1;
    } else {
        $host = "unix/";
        $port = $listen;
    }

    DEBUG "Starting to listen to $listen";

    return tcp_server $host, $port, sub { $self->_accept_handler($is_tcp, @_); }
}

sub _accept_handler {
    my ( $self, $is_tcp, $sock, $peer_host, $peer_port ) = @_;

    INFO "Accepted connection from $peer_host:$peer_port";

    return unless $sock;
    $self->{exit_guard}->begin;

    if ( $is_tcp && $self->{no_delay} ) {
        setsockopt($sock, IPPROTO_TCP, TCP_NODELAY, 1)
            or die "setsockopt(TCP_NODELAY) failed:$!";
    }

    my $handle = AnyEvent::Handle->new(
        fh        => $sock,
        timeout   => $self->{timeout},
        rtimeout  => $self->{timeout},
        wtimeout  => $self->{timeout},
        keepalive => $self->{keepalive},
    );
    my $ret = AE::cv;

    my $cleanup = sub {
        DEBUG "Closing connection";
        my $err = $_[2];
        $handle->destroy;
        if ($err) {
            ERROR "Connection error: $err";
            $ret->send($err);
        }
    };

    $handle->on_error($cleanup);
    $handle->on_eof($cleanup);
    $handle->on_timeout($cleanup);

    $handle->push_read(line => $self->_read_commands );
}

sub _read_commands
{
    my $self = shift;

    my $read_commands;
    $read_commands = sub {
        my ($handle, $cmd) = @_;

        DEBUG "Client> $cmd";

        if (
            my ($key, $flag, $ttl, $length) = 
                ($cmd =~ m/^set ([\p{IsWord}\/|:-]{1,250}) (\d{1,5}) (\d{1,11}) (\d{1,16})( noreply)?\s*$/)
        ) {
            $handle->push_read(chunk => $length, sub {
                my $data = $_[1];
                $self->storage->set($key, $flag, $ttl, \$data ) ?
                    $handle->push_write("STORED\r\n") :
                    $handle->push_write("NOT_STORED\r\n");
            } );
        }

        elsif (
            my ($keys_line) =
                ($cmd =~ m/^get (([\p{IsWord}\/|:-]{1,250})(\*([\p{IsWord}\/|:-]{1,250})){0,127})\s*$/)
        ) {
            my @keys = split /\*/, $keys_line, 128;
            my $items = $self->storage->get(\@keys);

            if ($items) {
                while (my ($item_key, $item) = each %$items) {
                    $handle->push_write("VALUE $item_key $item->[0] $item->[1]\r\n");
                    $handle->push_write("$item->[2]\r\n");
                }
                $handle->push_write("END\r\n");
            } else {
                $handle->push_write("ERROR\r\n");
            }
        }

        elsif (
            my ($delkey) = 
                ($cmd =~ m/^delete ([\p{IsWord}\/|:-]{1,250})(\s(\d{1,11}))?(\s(noreply))?\s*$/)
        ) {
            $self->storage->delete($delkey) ?
                $handle->push_write("DELETED\r\n") :
                $handle->push_write("NOT_FOUND\r\n");
        }

        elsif ($cmd =~ m/^stats items\s*$/) {
            my $number = $self->storage->stats_items_number;
            $handle->push_write("STAT items:1:number $number\r\n");
            $handle->push_write("END\r\n");
        }

        elsif (
            my ($slab, $limit) =
                ($cmd =~ m/^stats cachedump (\d{1,6}) (\d{1,6})\s*$/)
        ) {
            my $cachedump = $self->storage->stats_cachedump($slab, $limit);

            if ($cachedump) {
                while (my ($item_key, $item) = each %$cachedump) {
                    $handle->push_write("ITEM $item_key [$item->[0] b; $item->[1] s]\r\n");
                }
                $handle->push_write("END\r\n");
            } else {
                $handle->push_write("ERROR\r\n");
            }
        }

        elsif ($cmd =~ m/^version\s*$/) {
            $handle->push_write("VERSION $VERSION\r\n");
        }

        elsif ($cmd =~ m/^quit\s*$/) {
            $handle->destroy;
        }

        elsif ($cmd =~ m/\P{IsCntrl}/) {
            $handle->push_write("CLIENT_ERROR bad command line\r\n");
        }

        $handle->push_read(line => $read_commands);
    };

    return $read_commands;
}

1;
__END__

lib/MemcachedPP/Storage.pm
package MemcachedPP::Storage;

use strict;
use warnings;

use AnyEvent;


sub new
{
    my $class = shift;

    return bless {
        cache => {},
    }, ref $class || $class;
}

sub get
{
    my ($self, $keys) = @_;
    return 0 unless @$keys;
    my $items = {};
    my $now = AnyEvent->now;

    foreach my $key (@$keys) {
        if (exists $self->{cache}{$key}) {
            my $item = $self->{cache}{$key};

            $items->{$key} = [ $item->[0], length($item->[1]), $item->[1] ]
                if (($item->[0] == 0) || ($now < $item->[0]));
        }
    }

    return $items;
}

sub set
{
    my ($self, $key, $flag, $exptime, $value) = @_;
    return 0 unless length($key);
    $exptime ||= 0;
    $value ||= '';
    $exptime += AnyEvent->now if ((2592000 > $exptime) && ($exptime > 0));
    $exptime ++ if ($exptime % 2);
    $self->_set($key, int $exptime, $value);
    return 1;
}

sub _set
{
    my ($self, $key, $exptime, $value) = @_;
    $self->{cache}{$key} = [ $exptime, $$value ];
    return 1;
}

sub delete
{
    my ($self, $key) = @_;
    return delete $self->{cache}{$key} ? 1 : 0;
}

sub cleanup
{
    my $self = shift;
    my $now = AnyEvent->now;

    for (keys %{ $self->{cache} }) {
        $self->delete($_) if (
            ( $self->{cache}{$_}->[0] > 0 ) &&
            ( $self->{cache}{$_}->[0] <= $now )
        );
    }

    return 1;
}

sub stats_cachedump
{
    my ($self, $slab, $limit) = @_;
    my $items = {};
    my $cnt = 0;

    return $items if ( ($limit <= 0) or !$self->stats_items_number );

    while (
        ($cnt < $limit) &&
        ( my ($key, $item) = each %{ $self->{cache} } )
    ) {
        $items->{$key} = [ length($item->[1]), $item->[0] ];
        $cnt++;
    }

    return $items;
}

sub stats_items_number
{
    my $self = shift;
    $self->cleanup;
    return scalar keys %{ $self->{cache} };
}

1;
__END__

lib/MemcachedPP/Storage/SQLite.pm
package MemcachedPP::Storage::SQLite;

use strict;
use warnings;

use base qw(MemcachedPP::Storage);
use DBI qw(:sql_types);
require DBD::SQLite;

sub new
{
    my ($class, %args) = @_;
    $args{dbfile} ||= 'sqlite.db';

    my $self = bless \%args, ref $class || $class;
    $self->connect;

    return $self;
}

sub connect
{
    my $self = shift;

    my $dbh = DBI->connect(
        'dbi:SQLite:dbname=' . $self->{dbfile}, '', '',
        { RaiseError => 0, AutoCommit => 1 }
    )
        or die "Failed to connect to SQLite database within file '" .
            $self->{dbfile} . "': @!\n";

    $dbh->do( q{
        CREATE TABLE IF NOT EXISTS memcachedpp(
            id      VARCHAR(250) PRIMARY KEY ASC NOT NULL,
            exptime INTEGER,
            value   BLOB
        );
    } );

    $dbh->do( q{ DELETE FROM memcachedpp; } ) if $self->{reset};

    return $self->{dbh} = $dbh;
}

sub get
{
    my ($self, $keys) = @_;
    return 0 unless @$keys;
    my $items = {};

    my $sth = $self->{dbh}->prepare("
        SELECT id, exptime, value FROM memcachedpp WHERE
        id in (?" . ',?' x $#$keys . ") AND
        ( exptime = 0 OR exptime > strftime('%s','now') );
    ") or die @!;
    $sth->execute(@$keys);

    while (my $row = $sth->fetchrow_arrayref) {
        $items->{$row->[0]} = [ $row->[1], length($row->[2]), $row->[2] ];
    }

    return $items;
}

sub _set
{
    my ($self, $key, $exptime, $value) = @_;

    $self->delete($key);

    my $sth = $self->{dbh}->prepare( q{
        INSERT INTO memcachedpp (id, exptime, value) VALUES (?, ?, ?);
    } );
    $sth->bind_param(1, $key, SQL_VARCHAR);
    $sth->bind_param(2, $exptime, SQL_INTEGER);
    $sth->bind_param(3, $$value, SQL_BLOB);
    $sth->execute();

    return 1;
}

sub delete
{
    my ($self, $key) = @_;
    my $deleted = $self->{dbh}->do( q{
        DELETE FROM memcachedpp WHERE id = ?;
    }, undef, $key);

    return int $deleted;
}

sub cleanup
{
    my $self = shift;
    my $deleted = $self->{dbh}->do( q{
        DELETE FROM memcachedpp WHERE
        exptime != 0 AND exptime <= strftime('%s','now');
    } );

    return int $deleted;
}

sub stats_cachedump
{
    my ($self, $slab, $limit) = @_;
    my $items = {};
    return $items if ($limit <= 0);

    $self->cleanup;

    my $sth = $self->{dbh}->prepare( q{
        SELECT id, exptime, length(value) FROM memcachedpp WHERE
        exptime = 0 OR exptime > strftime('%s','now')
        LIMIT ?;
    } );
    $sth->execute($limit);

    while (my $row = $sth->fetchrow_arrayref) {
        $items->{$row->[0]} = [ int $row->[2], int $row->[1] ];
    }

    return $items;
}

sub stats_items_number
{
    my $self = shift;

    my $sth = $self->{dbh}->prepare( q{
        SELECT COUNT(id) FROM memcachedpp WHERE
        exptime = 0 OR exptime > strftime('%s','now');
    } );
    $sth->execute();

    return int $sth->fetchrow_arrayref->[0];
}

1;
__END__

Build.PL
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
    module_name         => 'MemcachedPP',
    license             => 'perl',
    dist_author         => 'Alexey Skorikov <alexey@skorikov.name>',
    dist_version_from   => 'lib/MemcachedPP.pm',
    dist_abstract       => 'README',
    requires => {
        'Log::Dispatch'       => 2.26,
        'Cache::Memcached'    => 1.29,
        'Mojolicious'         => 0.9999,
        'AnyEvent'            => 5.27,
        'EV'                  => 3.9,
        'DBI'                 => 1.61,
        'DBD::SQLite'         => 1.29,
        'Test::More'          => 0,
        'Test::Pod'           => 1.14,
        'File::Temp'          => 0,
        'version'             => 0,
    },
    add_to_cleanup      => [ 'MemcachedPP-*' ],
);

$builder->create_build_script();
